torch/csrc/dynamo/eval_frame.c

```diff
@@ -322,12 +343,150 @@ static void destroy_cache_entry(CacheEntry* e) {
   free(e);
 }
 
-inline static CacheEntry* get_cache_entry(PyCodeObject* code) {
-  CacheEntry* extra = NULL;
-  _PyCode_GetExtra((PyObject*)code, cache_entry_extra_index, (void*)&extra);
+/* CacheEntry helper functions ends */
+
+/* Extractions helper functions begins. They help with NULL and SKIP_CODE corner cases */
+
+inline static CacheEntry* extract_cache_entry(ExtraState* extra_state) {
+  // Helper to extra the cache_entry from the extra state.
+
+  // Ownership contract
+  // args
+  //  - extra_state: Borrowed
+  // return
+  //  - CacheEntry: Borrowed. set_extra_state is the owner.
+  if (extra_state == NULL || extra_state == SKIP_CODE) {
+    return NULL;
+  }
+  return extra_state->cache_entry;
+}
+
+
+inline static FrameState* extract_frame_state(ExtraState* extra_state) {
+  // Returns either the previously stored frame state or an empty dict.
+
+  // Ownership contract
+  // args
+  //  - extra_state: Borrowed
+  // return
+  //  - extra_state->frame_state: Borrowed. set_extra_state is the owner.
+  if (extra_state == NULL || extra_state == SKIP_CODE) {
+    return NULL;
+  }
+  return extra_state->frame_state;
+}
+
+/* Extractions helper functions ends */
+
+/* Extra state helper functions begins */
+
+inline static ExtraState* get_extra_state(PyCodeObject* code) {
+  // Ownership contract
+  // args
+  //  - code: Borrowed
+  // return
+  //  - extra_state: Borrowed. set_extra_state is the owner.
+  ExtraState* extra = NULL;
+  _PyCode_GetExtra((PyObject*)code, extra_index, (void*)&extra);
   return extra;
 }
 
+inline static void destroy_extra_state(PyCodeObject* code) {
+  // Destroys the extra state by deleting cache_entry, frame state and finally
+  // freeing the constructed extra state.
+
+  // Developer note - You should not call this function directly. This is called
+  // directly inside set_extra_state. If you are in a situation trying to call
+  // this function, consider if set_extra_state should be called.
+  ExtraState* extra = get_extra_state(code);
+  if (extra != NULL && extra != SKIP_CODE) {
+    CacheEntry* cache_entry = extract_cache_entry(extra);
+    FrameState* frame_state = extract_frame_state(extra);
+    destroy_cache_entry(cache_entry);
+    Py_XDECREF(frame_state);
+    free(extra);
+  }
+}
+
+inline static void set_extra_state(PyCodeObject* code, ExtraState* extra_state) {
+  // Clears the existing object sitting on the extra scratch spance and sets it
+  // up with the new state.
+
+  // Ownership contract
+  // args
+  //  - extra_state: Stolen
+  // return
+  //  - there is no return, but the extra_state is stolen, so it becomes
+  //  set_extra_state responsibility to clean it up. It will be deleled during
+  //  the reset_code/skip, when the set_extra_state is called with
+  //  NULL/SKIP_CODE.
```

Comment:
There is another invariant, which is you must not try to set extra state for an extra state that is ALREADY on that code object. You will destroy the extra state and then write something invalid into SetExtra. We could test this invariant by running `PyCode_GetExtra` first, if you don't mind the extra CPython API call (it might be fine, I'm not sure.)

Reply:
Added. Its fine, its not in the critical path. This code will get triggered only when we are recompiling.